CONTRIBUTING.md

```diff
@@ -252,6 +252,74 @@ There may, at times, need to be exceptions where breaking changes are allowed in
 discretion of the project's maintainers, and must be carefully considered before merging. An example of an allowed
 breaking change might be a fix for a behavioral bug that was released in an initial minor version (such as `v0.3.0`).
 
+
+## API conventions
+
+This project follows the [Kubernetes API conventions](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/api-conventions.md). Minor modifications or additions to the conventions are listed below.
+
+### Optional vs. Required
```

Comment:
can we mention that fields that are conceptually required when are pointers and defaulted are marked as //+optional because of a technical limitation. i.e CRDs don't publish defaults, so client side validation (e.g kubectl) breaks if the field is given empty input because it does not know that it would be defaulted on the server.

Reply:
Yes. I assume you're referring to cases like this? https://github.com/kubernetes-sigs/cluster-api/pull/4527
(I'll then use it as example)

Reply:
I would add the following:
> * If a field is required (for our controllers to work) and has a default value, but we don't want to force users to set the field, we have to 
>  mark the field as optional. Otherwise the client-side kubectl OpenAPI schema validation will force the user to set it even though it would be
>  defaulted on the server-side.

I'm not entirely sure about how we should make the field optional. I think I would only add the `+optonal` annotation as that's the closes to how a required field should look like. This is not necessary consistent with our current fields (e.g. `MachineDeployment.Spec.MachineDeploymentStrategy`)

Any thoughts? (@enxebre @vincepri @fabriziopandini)

Reply:
Can we add this in a follow up PR (same as printing status vs intent) ?

Reply:
Moved to https://github.com/kubernetes-sigs/cluster-api/issues/5333